src/NumericalAlgorithms/Interpolation/MultiLinearSpanInterpolation.hpp

```diff
@@ -0,0 +1,397 @@
+  DataPointer y;
+
+ public:
+  template <size_t thisDimension>
+  struct weight_t {
+    std::array<
+        double,
+        brigand::power<std::integral_constant<size_t, 2>,
+                       std::integral_constant<size_t, thisDimension>>::value>
+        weights;
+    Index<brigand::power<std::integral_constant<size_t, 2>,
+                         std::integral_constant<size_t, thisDimension>>::value>
+        index;
+  };
+
+  size_t extents(size_t const whichDimension) const {
+    return numberOfPoints[whichDimension];
+  }
+
+  void extrapolateAboveData(size_t whichDimension, bool value) {
+    allow_extrapolation_above_data[whichDimension] = value;
+  }
+
+  void extrapolateBelowData(size_t whichDimension, bool value) {
+    allow_extrapolation_below_data[whichDimension] = value;
+  }
+
+  /// Determine relative index bracketing for non-uniform spacing
+  size_t findIndexGeneral(const size_t whichDimension,
+                          const double& targetPoints) const;
+
+  //! Determine relative index bracketing for non-uniform spacing
+  size_t findIndexUniform(const size_t whichDimension,
+                          const double& targetPoints) const;
+
+  size_t findIndex(const size_t whichDimension,
+                   const double& targetPoints) const {
+    if constexpr (UniformSpacing) {
+      return findIndexUniform(whichDimension, targetPoints);
+    } else {
+      return findIndexGeneral(whichDimension, targetPoints);
+    }
+  };
+
+  /// Compute interpolation weights for 1D tables
+  weight_t<1> getWeights(const double x1) const;
+
+  /// Compute interpolation weights for 2D tables
+  weight_t<2> getWeights(const double x1, const double x2) const;
+
+  /// Compute interpolation weights for 3D tables
+  weight_t<3> getWeights(const double x1, const double x2,
+                         const double x3) const;
+
+  double interpolate(weight_t<Dimension> const& weights,
+                     size_t whichVariable = 0) const {
+    double result = 0;
+
+    for (size_t nn = 0; nn < weights.weights.size(); ++nn)
+      result += weights.weights[nn] *
+                y[whichVariable + NumberOfVariables * weights.index[nn]];
+
+    return result;
+  };
+
+  template <size_t... VariablesToInterpolate>
+  std::array<double, sizeof...(VariablesToInterpolate)> interpolate(
+      weight_t<Dimension> const& weights) const {
+    static_assert(sizeof...(VariablesToInterpolate) <= NumberOfVariables,
+                  "You are trying to interpolate more variables than this "
+                  "container holds.");
+
+    return std::array<double, sizeof...(VariablesToInterpolate)>{
+        interpolate(weights, VariablesToInterpolate)...};
+  }
+
+  template <size_t NumberOfVariablesToInterpolate, typename... T,
+            Requires<(std::is_floating_point_v<typename std::remove_cv_t<T>> and
+                      ...)> = nullptr>
+  std::array<double, NumberOfVariablesToInterpolate> interpolate(
```

Comment:
Question: what do you think of always forcing people to use the function that takes `get_weights()`? That way people are more likely to reuse weights :) So basically have calls be `interpolator.intperpolate<Vars...>(interpolator.get_weights(x0, x1, x2))`? It's more typing in some cases, but will hopefully encourage reuse...

Reply:
I'm not sure. What happens if we only store one variable? 
Then the weights would not be useful... 

Is there a policy on using stuff like [[deprecated]] etc.?

Reply:
Okay, sounds good! Let's try this and once we have some experience using the code we can always adjust :) 

We can use `[[deprecated]]` but we also use `-Werror` in CI so I'm not sure if that becomes problematic. I guess not because it would just prevent folks from committing new code that uses it.

Reply:
OK!